src/nbl/ui/CWindowWin32.cpp

```diff
@@ -9,16 +11,21 @@ namespace nbl {
 namespace ui
 {
 
-	CWindowWin32::CWindowWin32(core::smart_refctd_ptr<system::ISystem>&& sys, int _x, int _y, uint32_t _w, uint32_t _h, E_CREATE_FLAGS _flags)
+	CWindowWin32::CWindowWin32(CWindowManagerWin32* winManager, core::smart_refctd_ptr<system::ISystem>&& sys, int _x, int _y, uint32_t _w, uint32_t _h, E_CREATE_FLAGS _flags, const std::string_view& caption) : m_windowManager(winManager)
 	{
 		m_width = _w; m_height = _h;
-		windowThreadHandler.createWindow(_x, _y, _w, _h, _flags, m_native);
+		m_native = m_windowManager->createNativeWindow(_x, _y, _w, _h, _flags, caption);
```

Comment:
this is really roundabout, I'd much rather prefer a constructor that takes a native window and the window manager

Reply:
because we dont want to use exceptions, so the only way to have a "constructor which can fail" is to have a private "constructor that cannot fail" and either a friended factory class OR a static create method, which does all the setup and validation before calling the ctor.

Reply:
window manager should have `createWindow()` method which returns `smart_refctd_ptr<IWindow>`. Already commented about this, a while ago

Reply:
`createWindow()` should validate and return nullptr in case of failed validation and if native handle creation failed

Reply:
> `createWindow()` should validate and return nullptr in case of failed validation and if native handle creation failed

yep so thats stuff that happens before calling a constructor on the window class